include/dlaf/factorization/qr/t_factor_mc.h

```diff
@@ -0,0 +1,191 @@
+  const auto t_size = t.tileSize(t_idx);
+  DLAF_ASSERT(k <= t_size.rows(), k, t_size);
+  DLAF_ASSERT(k <= t_size.cols(), k, t_size);
+
+  if (rank.col() != rank_v0.col())
+    return;
+
+  const LocalTileIndex v_start_loc{
+      dist.template nextLocalTileFromGlobalTile<Coord::Row>(v_start.row()),
+      dist.template nextLocalTileFromGlobalTile<Coord::Col>(v_start.col()),
+  };
+  const LocalTileIndex v_end_loc{dist.localNrTiles().rows(), v_start_loc.col() + 1};
+
+  t(t_idx).then(unwrapping([](auto&& tile) {
+    lapack::laset(lapack::MatrixType::General, tile.size().rows(), tile.size().cols(), 0, 0, tile.ptr(),
+                  tile.ld());
+  }));
+
+  // T(0:j, j) = T(0:j, 0:j) . -tau(j) . V(j:, 0:j)* . V(j:, j)
+  for (SizeType j = 0; j < k; ++j) {
+    // this is the x0 element of the reflector j
+    const TileElementIndex x0{j, j};
+
+    const TileElementIndex t_start{0, x0.col()};
+    const TileElementSize t_size{x0.row(), 1};
+
+    // 2A First step GEMV
+    for (const auto& v_i_loc : iterate_range2d(v_start_loc, v_end_loc)) {
+      const SizeType v_i = dist.template globalTileFromLocalTile<Coord::Row>(v_i_loc.row());
+
+      const bool is_v0 = (v_i == v_start.row());
+
+      auto gemv_func = unwrapping([=](const auto& tile_v, const T tau, auto&& tile_t) {
+        const SizeType first_element_in_tile = is_v0 ? x0.row() : 0;
+
+        // T(0:j, j) = -tau . V(j:, 0:j)* . V(j:, j)
+        // [j x 1] = [(n-j) x j]* . [(n-j) x 1]
+        TileElementSize va_size{tile_v.size().rows() - first_element_in_tile, x0.col()};
+        TileElementIndex va_start{first_element_in_tile, 0};
+        TileElementIndex vb_start{first_element_in_tile, x0.col()};
+
+        if (is_v0) {
+          tile_t(x0) = tau;
+
+          // use implicit 1 for the 2nd operand
+          for (SizeType r = 0; !va_size.isEmpty() && r < va_size.cols(); ++r) {
+            const TileElementIndex i_v{va_start.row(), r + va_start.col()};
+            const TileElementIndex i_t{r + t_start.row(), t_start.col()};
+
+            DLAF_ASSERT_HEAVY(i_t.isIn(tile_t.size()), i_t, t_size);
+            DLAF_ASSERT_HEAVY(i_v.isIn(tile_v.size()), i_v, tile_v.size());
+
+            tile_t(i_t) = -tau * dlaf::conj(tile_v(i_v));
+          }
+
+          // skip alredy managed computations with implicit 1
+          va_start = va_start + TileElementSize{1, 0};
+          vb_start = vb_start + TileElementSize{1, 0};
+          va_size = {va_size.rows() - 1, va_size.cols()};
+        }
+
+        if (!va_size.isEmpty()) {
+          // clang-format off
+          blas::gemv(blas::Layout::ColMajor,
+              blas::Op::ConjTrans,
+              va_size.rows(), va_size.cols(),
+              -tau,
+              tile_v.ptr(va_start), tile_v.ld(),
+              tile_v.ptr(vb_start), 1,
+              1, tile_t.ptr(t_start), 1);
+          // clang-format on
+        }
+      });
+
+      hpx::dataflow(gemv_func, v.read(v_i_loc), taus[j], t(t_idx));
```

Comment:
I see a problem here: All the gemv are serialized as all need to write to the same t.

1st solution (easy): keep the gemv serialized, but exchange the j and v_i_loc the loop and put the j loop in the task (such that the V tile is read only once in cache).
2nd solution: 1st solution + have multiple T and reduce them before all_reduce.

Reply:
Good point. I followed your first suggestion for the moment, for the second I left there a comment so we can improve it in the future.